include/seahorn/Expr/ExprRewriter.hh

```diff
@@ -0,0 +1,144 @@
+#pragma once
+#include "seahorn/Expr/Expr.hh"
+#include "seahorn/Expr/ExprCore.hh"
+#include "seahorn/Expr/ExprRewriteRule.hh"
+
+namespace expr {
+using namespace seahorn;
+
+Expr rewriteITEComp(Expr exp);
+
+struct RewriteFrame {
+  Expr m_exp;
+  size_t m_depth;
+  size_t m_i; // up to m_i th children have been rewritten
+  RewriteFrame(Expr exp, size_t depth, size_t i)
+      : m_exp(exp), m_depth(depth), m_i(i) {}
+};
+
+using ExprVisitedMap = std::unordered_map<ENode *, bool>;
+using RewriteFrameVector = std::vector<RewriteFrame>;
+
+class ExprRewriterConfig {
+  /* apply rewrite rules */
+protected:
+  ExprFactory &m_efac; // for making expr
+  EZ3 &m_zctx;         // for z3 simplifier if needed
+
+public:
+  ExprRewriterConfig(ExprFactory &efac, EZ3 &zctx)
+      : m_efac(efac), m_zctx(zctx) {}
+
+  rewrite_result applyRewriteRules(Expr exp);
+
+  bool shouldRewrite(Expr exp);
+};
+
+/* example config for ITE */
+class ITECompRewriteConfig : public ExprRewriterConfig {
+private:
+  ITERewriteRule m_iteRule;
+  CompareRewriteRule m_compRule;
+
+public:
+  ITECompRewriteConfig(ExprFactory &efac, EZ3 &zctx)
+      : m_iteRule(efac, zctx), m_compRule(efac, zctx),
+        ExprRewriterConfig(efac, zctx) {}
+
+  rewrite_result applyRewriteRules(Expr exp);
+
+  bool shouldRewrite(Expr exp);
+};
+
+template <typename Config> class ExprRewriter {
+protected:
+  Config m_config;
+  ExprFactory &m_efac; // for making expr
+  EZ3 &m_zctx;         // for z3 simplifier if needed
+
+  RewriteFrameVector m_rewriteStack;
+  ExprVector m_resultStack;
+  DagVisitCache m_cache;
+  ExprVisitedMap m_visited;
+
+  /* visit e, return true if any of the following is true:
+    1. e has been cached, or
+    2. current depth is 0 or
+    3. current m_config is dictates e should not be further visited
+    in any of the above case, push e to top of result stack
+    otherwise return false and push e into top of rewriteStack
+  */
+  bool visited(Expr e, size_t depth) {
```

Comment:
the function has too many side-effects. The name does not reflect them all, neither does the comment.

Reply:
perhaps renaming the function into `visit` would go long way to address the comment above